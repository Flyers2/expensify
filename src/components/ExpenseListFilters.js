import React from 'react';
import {connect} from 'react-redux';
import filters from "../reducers/filters";
import {setStartDate, setTextFilter, sortByAmount, sortByDate, setEndDate} from "../actions/filters";
import {DateRangePicker} from 'react-dates';


export class ExpenseListFilters extends React.Component {

    state = {
        calendarFocused: null
    };

    onDatesChange = ({startDate, endDate}) => {
        this.props.mappedSetStartDate(startDate);
        this.props.mappedSetEndDate(endDate);
    };
    onFocusChange = (calendarFocused) => {
        this.setState(() => ({calendarFocused}))
    };

    changeFilter = (e) => {
     //   console.log(this.props + "here in change filter")
        this.props.mappedSetTextFilter(e.target.value);

    };
    onSortChange = (e) =>  {

        const sortType = e.target.value;
        if (sortType === 'date') {
            this.props.mappedSortByDate();
        }
        if (sortType === 'amount') {
            this.props.mappedSortByAmount();
        }
    };


    render() {
        return (

            <div>
                <input type="text" className='filterInput' value={this.props.filters.text} onChange={this.changeFilter}/>

                <select onChange={this.onSortChange}>
                    <option value="date">Date</option>
                    <option value="amount">Amount</option>
                </select>

                <DateRangePicker
                    startDate={this.props.filters.startDate}
                    endDate={this.props.filters.endDate}
                    onDatesChange={this.onDatesChange}
                    focusedInput={this.state.calendarFocused}
                    onFocusChange={this.onFocusChange}
                    numberOfMonths={1}
                    showClearDates={true}
                    isOutsideRange={() => false}

                />

            </div>
        )


    }

}


const mapStateToProps = (state) => {
    return (
        {filters: state.filters}
    )

};

const mapDispatchToProps=(dispatch)=>({
    mappedSetTextFilter:(text)=>dispatch(setTextFilter(text)),
    mappedSortByDate:()=>dispatch(sortByDate()),
    mappedSortByAmount:()=>dispatch(sortByAmount()),
    mappedSetStartDate:(date)=>dispatch(setStartDate(date)),
    mappedSetEndDate:(date)=>dispatch(setEndDate(date))

});
export default connect(mapStateToProps,mapDispatchToProps)(ExpenseListFilters);