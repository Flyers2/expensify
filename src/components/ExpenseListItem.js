import React from 'react';


import {Link} from 'react-router-dom'


export const ExpenseListItem = (props)=>{

    return(
        <div>
            <Link to={`/edit/${props.id}`}>
                <h3>{props.description} </h3>
            </Link> - {props.amount}
           -

        </div>

    )
};

const ConnectedExpenseListItem = ExpenseListItem
export default ConnectedExpenseListItem