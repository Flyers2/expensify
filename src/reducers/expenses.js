



const expensesDefaultState = [];
export default (state = expensesDefaultState, action) => {

    switch (action.type) {
        case 'ADD_EXPENSE':
            return [...state, action.expense];
        case 'REMOVE_EXPENSE':
            const id = action.id;
         //   console.log(action)
        //    console.log(`removing epense with id  of ${id}`);
            return state.filter((expense)=>expense.id!==id);

        case 'EDIT_EXPENSE':
            return state.map((expense)=>{
               // console.log(`expense id in loop is ${expense.id} -- looking for ${action.id}`)
                if(expense.id==action.id){
        //            console.log(action.updates)
                    return{
                        ...expense,
                        ...action.updates
                    }
                }else{
                  //  console.log(`expense reducer did not find a matching id`)
                    return expense;
                }
            });

        default :
            return state;
    }
};