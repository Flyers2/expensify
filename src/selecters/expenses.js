import moment from 'moment';


export default (expenses,{text,sortBy,startDate, endDate})=>{


    return expenses.filter((expense)=>{
    //    console.log(expense)
        const createdAtMoment = moment(expense.createdAt);
      //  const startDateMatch = typeof startDate !=='number'|| expense.createdAt>= startDate;
        const startDateMatch = startDate ? startDate.isSameOrBefore(createdAtMoment,'day') : true
        const endDateMatch = endDate? endDate.isSameOrAfter(createdAtMoment,'day'): true;
        const textMatch = typeof text!=='string' || expense.description.toLowerCase().includes(text.toLowerCase());

        return startDateMatch && endDateMatch &&  textMatch;
    }).sort((expense1,expense2,)=>{

        if(sortBy==='date'){
          //  console.log(`checking which date is bigger ${expense1.createdAt}vs ${expense2.createdAt}`)
            return expense1.createdAt< expense2.createdAt?1:-1;
        }if(sortBy==='amount'){
            return expense1.amount>expense2.amount?-1:1;
        }
    });
}
