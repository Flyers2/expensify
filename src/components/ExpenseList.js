import React from 'react';
import {connect} from 'react-redux';
import ExpenseListItem from './ExpenseListItem'
import selectExpenses from '../selecters/expenses';

export const ExpenseList = (props) => (
    <div>

        {props.expenses.length ===0 &&
            <p>No Expenses</p>
        }

        {props.expenses.length > 0 &&
        props.expenses.map((exp, i) => {
                return (
                    <ExpenseListItem
                        key={exp.id}
                        {...exp}
                    />

                )
            }
        )

        }
    </div>
);


const mapStateToProps = (state) => {
    return {
        expenses: selectExpenses(state.expenses, state.filters)
    }
};

const ConnectedExpenseLIst = connect(mapStateToProps)(ExpenseList);

export default ConnectedExpenseLIst;


