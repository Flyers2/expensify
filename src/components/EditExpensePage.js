import React from 'react';
import {connect} from 'react-redux'
import ExpenseForm from "./ExpenseForm";
import {editExpense, removeExpense} from "../actions/expenses";





export class EditExpensePage extends React.Component {

    onSubmit = (newExpense) => {
           // console.log('submitting expense:',newExpense)
           // console.log('submitting expense id:',this.props.expense.id)
        this.props.outerEditExpense( this.props.expense.id,newExpense);
        //      outerEditExpense(this.props.match.params.id, expense);
        this.props.history.push('/')
    }

    deleteExpense = () => {
     //   console.log(this.props, 'from delete expense function in editExpense.js')
        this.props.outerDeleteExpense(this.props.expense.id);
        this.props.history.push('/')
    }

    render() {
        return (

            <div>
                <ExpenseForm expense={this.props.expense} onSubmit={this.onSubmit}/>

                <p>
                    <button className='removeExpense' onClick={ this.deleteExpense} >Remove</button>
                </p>


            </div>
        )


    }
}

const mapStateToProps = (state, props) => {
    const foundex=state.expenses.find((ex) => ex.id === props.match.params.id);
   // console.log(foundex)
    return {
        expense:foundex ,
        fakeprop: "am i here?"
    }

}
const mapDispatchToProps = (dispatch) => {
    return {
        outerEditExpense: ( id,expense) => {
            dispatch(editExpense(id, expense));
        },
        outerDeleteExpense:(id)=>(dispatch(removeExpense(id)))
    }
}

export default connect(mapStateToProps, mapDispatchToProps)(EditExpensePage);