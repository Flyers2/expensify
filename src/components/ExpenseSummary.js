import React from 'react';
import {connect} from 'react-redux'
import {getTotal} from "../selecters/expenses-total";
import numeral from 'numeral';
import selectExpenses from '../selecters/expenses';


export const ExpenseSummary = (props) => {
    return (
        props.numberOfExpenses>0 ?
            <div>There are {props.numberOfExpenses} visible Expenses for a total of { numeral(props.totalExpenses/100).format('$0,0.00')}</div>
            :
          "There are no visible Expenses"

    )
};

const mapStateToProps = (state, props) => {
    return {
        totalExpenses: getTotal(selectExpenses( state.expenses,state.filters)),
        numberOfExpenses:selectExpenses( state.expenses,state.filters).length
    }
}


export default connect(mapStateToProps)(ExpenseSummary);


//  <div>There are {props.numberOfExpenses} Expenses for a total of { numeral(props.totalExpenses).format('$0,0.00')}</div>