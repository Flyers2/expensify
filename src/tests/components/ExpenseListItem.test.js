import React from 'react';
import {shallow} from 'enzyme'
import expenses from '../fixtures/expenses';
import {ExpenseListItem} from '../../components/ExpenseListItem';

test('should render an expenseListItem',()=>{
    const wrapper = shallow(<ExpenseListItem
                             id={expenses[0].id}
                             description = {expenses[0].description}
                             amount = {expenses[0].amount}
                            />)
    expect(wrapper).toMatchSnapshot();
})