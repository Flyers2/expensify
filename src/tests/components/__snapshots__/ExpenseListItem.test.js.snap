// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`should render an expenseListItem 1`] = `
<div>
  <Link
    replace={false}
    to="/edit/1"
  >
    <h3>
      Gum
       
    </h3>
  </Link>
   - 
  195
  -
</div>
`;
